### PR TITLE
Add full ExchangeWorkflow E2E test

### PR DESCRIPTION
**Pull request by efoxepstein**
This fixes some major bugs in `panel-exchange-client`, completely rewrites how Apache Beam ExchangeTasks are handled, and adds better test coverage.

This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220)

### CODE REVIEW

File: src/test/kotlin/org/wfanet/panelmatch/integration/FullWorkflowTest.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  override val initialModelProviderInputs: Map<String, ByteString> =
+    mapOf("mp-plaintext-join-keys" to PLAINTEXT_JOIN_KEYS.toByteString())
+
+  override fun validateFinalState(
+    dataProviderDaemon: ExchangeWorkflowDaemonForTest,
+    modelProviderDaemon: ExchangeWorkflowDaemonForTest
+  ) {
+    val blob = modelProviderDaemon.readPrivateBlob("mp-decrypted-join-keys")
+    Truth.assertThat(blob).isNotNull()
```
Comment (yunyeng) [inline src/test/kotlin/org/wfanet/panelmatch/integration/FullWorkflowTest.kt:59]: In the Abstract Test class can we also check the steps size from service same as workflow steps size? Right before we assert the steps' states.

assertThat(steps.size).isEqualTo(exchangeWorkflow.stepsCount)

**Review by efoxepstein — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220)* status: 0 of 32 files reviewed, 1 unresolved discussion (waiting on @yunyeng)

___
*[src/test/kotlin/org/wfanet/panelmatch/integration/FullWorkflowTest.kt, line 59 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpCJG40hpWjwOQALSWR-r1-59:-MpDBsYV4vjxhEna_IVl:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/225fd4ebe5337330e71735dbb66aed0ea43fdc33/src/test/kotlin/org/wfanet/panelmatch/integration/FullWorkflowTest.kt#L59)):*
<details><summary><i>Previously, yunyeng (Yunus Yen) wrote…</i></summary><blockquote>

In the Abstract Test class can we also check the steps size from service same as workflow steps size? Right before we assert the steps' states.

assertThat(steps.size).isEqualTo(exchangeWorkflow.stepsCount)
</blockquote></details>

Done.

**Review by duliomatos1 — APPROVED** [review]
Looks good to me, some optional suggestions that shouldn't block the review

> Reviewed 2 of 5 files at r1, 30 of 30 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-:-MpHcTzV5zvfXXokVsWW:b-hpv2ah)* status: all files reviewed, 7 unresolved discussions (waiting on @efoxepstein and @yunyeng)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/privatemembership/WriteShardedData.kt, line 69 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpH_kk20M-EmpMDSvmt:-MpH_kk38BgaVu-ThiEs:bur9gri) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/client/privatemembership/WriteShardedData.kt#L69)):*
> ```Kotlin
>       input
>         .keyBy("Key by Blob") { it.hashCode() % shardCount }
>         //        .setCoder(KvCoder.of(VarIntCoder.of(), ProtoCoder.of()))
> ```

Are you keeping this commented out or was it just for testing something?
___
*[src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt, line 143 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpHb6H9-40Lt3bjMsZA:-MpHb6H9-40Lt3bjMsZB:bh9ag9k) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt#L143)):*
> ```Kotlin
> 
>     // Assert that we are not deadlocked:
>     val terminalStates = setOf(ExchangeStep.State.SUCCEEDED, ExchangeStep.State.FAILED)
> ```

this can be a val e.g. TERMINAL_STATES declared toplevel
___
*[src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt, line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpHbXiW14FGr62juAVd:-MpHbXiW14FGr62juAVe:b1arubl) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt#L145)):*
> ```Kotlin
>     val terminalStates = setOf(ExchangeStep.State.SUCCEEDED, ExchangeStep.State.FAILED)
>     if (steps.any { it.state !in terminalStates }) {
>       val readyStates =
> ```

same here, though it's not reused out of here
___
*[src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt, line 153 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpHajwo51K4YWNQ1TPp:-MpHajwo51K4YWNQ1TPq:b4mzzjs) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt#L153)):*
<details><summary><i alt="summary">Quoted 12 lines of code…</i></summary>

> ```
>     // Assert that we are not deadlocked:
>     val terminalStates = setOf(ExchangeStep.State.SUCCEEDED, ExchangeStep.State.FAILED)
>     if (steps.any { it.state !in terminalStates }) {
>       val readyStates =
>         setOf(
>           ExchangeStep.State.IN_PROGRESS,
>           ExchangeStep.State.READY,
>           ExchangeStep.State.READY_FOR_RETRY
>         )
> 
>       assertThat(steps.any { it.state in readyStates }).isTrue()
>     }
> ```
</details>

 This could be a separate function like `assertNotDeadlocked()`
___
*[src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt, line 165 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpHbpmIEt2dIgcrp0kS:-MpHbpmIEt2dIgcrp0kT:b47v56i) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt#L165)):*
per previous suggestion, could be
> ```
>     logStepStates()
>     assertNotDeadlocked()
> ```
___
*[src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt, line 167 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpHbH_g5XfI5F8YSnvV:-MpHbH_g5XfI5F8YSnvW:b-b3w6ur) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt#L167)):*
> ```Kotlin
>       logStepStates()
>       logger.info("Exchange is in state: ${exchange.state}.")
>       exchange.state in setOf(Exchange.State.SUCCEEDED, Exchange.State.FAILED)
> ```

could be `exchange.state in TERMINAL_STATES` per previous suggestion

**Review by yunyeng — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220)* status: all files reviewed, 6 unresolved discussions (waiting on @efoxepstein)

**Review by efoxepstein — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220)* status: all files reviewed, 6 unresolved discussions (waiting on @duliomatos1)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/privatemembership/WriteShardedData.kt, line 69 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpH_kk20M-EmpMDSvmt:-MpHkPSe7LKit_BkkS9n:bhtr4hk) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/client/privatemembership/WriteShardedData.kt#L69)):*
<details><summary><i>Previously, duliomatos1 (Dulio Matos) wrote…</i></summary><blockquote>

Are you keeping this commented out or was it just for testing something?
</blockquote></details>

Oops! I'll delete this
___
*[src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt, line 143 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpHb6H9-40Lt3bjMsZA:-MpHkkrdEjpltBCIxeBf:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt#L143)):*
<details><summary><i>Previously, duliomatos1 (Dulio Matos) wrote…</i></summary><blockquote>

this can be a val e.g. TERMINAL_STATES declared toplevel
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt, line 145 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpHbXiW14FGr62juAVd:-MpHklWx9E-fRCv89PE7:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt#L145)):*
<details><summary><i>Previously, duliomatos1 (Dulio Matos) wrote…</i></summary><blockquote>

same here, though it's not reused out of here
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt, line 153 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpHajwo51K4YWNQ1TPp:-MpHlD9jBDsz3b6LuLJc:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt#L153)):*
<details><summary><i>Previously, duliomatos1 (Dulio Matos) wrote…</i></summary><blockquote>

> ```
>     // Assert that we are not deadlocked:
>     val terminalStates = setOf(ExchangeStep.State.SUCCEEDED, ExchangeStep.State.FAILED)
>     if (steps.any { it.state !in terminalStates }) {
>       val readyStates =
>         setOf(
>           ExchangeStep.State.IN_PROGRESS,
>           ExchangeStep.State.READY,
>           ExchangeStep.State.READY_FOR_RETRY
>         )
> 
>       assertThat(steps.any { it.state in readyStates }).isTrue()
>     }
> ```
 This could be a separate function like `assertNotDeadlocked()`
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt, line 165 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpHbpmIEt2dIgcrp0kS:-MpHlE3h8DqZKlYRsih7:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt#L165)):*
<details><summary><i>Previously, duliomatos1 (Dulio Matos) wrote…</i></summary><blockquote>

per previous suggestion, could be
> ```
>     logStepStates()
>     assertNotDeadlocked()
> ```
</blockquote></details>

Done.
___
*[src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt, line 167 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/220#-MpHbH_g5XfI5F8YSnvV:-MpHlPAS-wlcVdcKHoPM:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/1185c4d71acf47b6e454dcd38a5bb8b144559644/src/main/kotlin/org/wfanet/panelmatch/integration/AbstractInProcessPanelMatchIntegrationTest.kt#L167)):*
<details><summary><i>Previously, duliomatos1 (Dulio Matos) wrote…</i></summary><blockquote>

could be `exchange.state in TERMINAL_STATES` per previous suggestion
</blockquote></details>

Done.

**Pull request merged**